Fixed disappearing email form bug by forcing a refresh after submission, Fixes #11565

MeisterRed opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #11565 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
This pull request resolves the issue involving the user changing the email address for their account. When the user filled out the form and clicked submit to change their email address, it originally updated the email for the account but also removed the "Change Email Address" and "Change Password" forms from the settings (acting as if the account didn't have an email). This caused confusion for users since they didn't know if the email for their account was actually updated until they refreshed the page.

The solution I created was to force the page to refresh after the user fills out the form and clicks submit. This will avoid causing confusion for the user


[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 7d427213-9991-49fb-8513-77e1456da77e

MeisterRed committed:
Fixed disappearing form bug by forcing a refresh after submission

MeisterRed committed:
Added reloadPage() to the changeEmail submit button

MeisterRed committed:
Added indentation for the reloadPage() function

paglias commented:
Thanks @MeisterRed ! There is a problem in the code because the `changeUser` function is an async one so `reloadPage` will be called immediately but the request to the server might not yet have completed which might cause problems.

Also, would you be open to try a fix where the page is not fully reloaded but instead the buttons don't disappear? Thanks

MeisterRed commented:
Alright, I'll see if I can make a few modifications so the refresh happens after the request is sent to the server.

I originally tried to find a fix that doesn't involve refreshing the whole page but I couldn't think of a proper way of preventing the form from disappearing. if you wish, I can try my best and try a fix within those requirements but I can't promise anything.

paglias commented:
> I originally tried to find a fix that doesn't involve refreshing the whole page but I couldn't think of a proper way of preventing the form from disappearing. if you wish, I can try my best and try a fix within those requirements but I can't promise anything.

No worries, if you want to try that would be great and I'm happy to answer any question or doubt you might have but if you don't think it's possible the full refresh will work for now

MeisterRed commented:
Thank you very much and sorry for for the late response. I'll fix up the code soon

MeisterRed committed:
Adjusted changeUser() function (added reloadPage)

Added reloadPage() function to the changeUser() function. This should allow the page to reload after the email is updated instead of before the function gets to finish updating the account's email

MeisterRed commented:
Alright, I adjusted the code a bit by adding window.location.reload inside of the changeUser function instead of having it run side by side with the changeUser function. I'm hoping that this will allow the email to update before the page refreshes entirely. 

If there's anything that needs to be fixed or adjusted, please let me know

MeisterRed committed:
Adjusted changeUser() function

Added  window.location.reload() to the changeUser() function. This should allow the page to reload after the email is updated instead of before the function gets to finish updating the account's email

paglias commented:
@MeisterRed so i looked at the code and while this fix works it would be great to avoid having to reload the whole page.

The reason this bug is happening is that this line https://github.com/HabitRPG/habitica/blob/0935c6e28edca4caed8054cb04b919f81e8816ab/website/client/src/components/settings/site.vue#L815 replaces the email value (used to decide if the `Change Email` part is displayed) to the result of the response from the server.

But the server doesn't return an attribute named `email` but `newEmail` so the email is set to undefined and the form disappears.

It can be fixed by changing that line to 

```
      } else if (attribute === 'email') {
        this.user.auth.local.email = updates.newEmail;
```

Let me know if you have any question

MeisterRed committed:
Adjusted changeUser function

MeisterRed commented:
I've made the changes you have suggested and that solved the bug without having to refresh the page. I also added a popup window that tells the user that the email has been successfully changed so they're not left hanging after clicking submit.
![image](https://user-images.githubusercontent.com/32404189/70943587-385e4a80-201f-11ea-8e0e-b93b0eb00de3.png)

paglias commented:
@MeisterRed  thanks for this! I'm merging the PR and awarding your contributor tier.

Note that further tiers require increasing amounts of work from one to the next, so Tier 2 may require a couple of PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!
